nutils/solver.py

```diff
@@ -605,29 +605,36 @@ def __init__(self, target:types.strictstr, residual:sample.strictintegral, inert
     self.inertia = inertia
     self.theta = theta
     self.timestep = timestep
+    self.timetarget = timetarget or '_thetamethod_dummy'
+    self.time0 = time0
 
   def _res_jac(self, timestep):
-    res = self.residual * self.theta + self.inertia / timestep \
+    res = (self.residual * self.theta + self.inertia / timestep).replace({self.timetarget: function.Argument(self.timetarget, ())+timestep}) \
         + (self.residual * (1-self.theta) - self.inertia / timestep).replace({self.target: function.Argument(self.target0, self.lhs0.shape)})
     return res, res.derivative(self.target)
 
-  def _step(self, lhs, timestep):
+  def _step(self, lhs, t, timestep):
     res, jac = self._res_jac(timestep)
     try:
       return newton(self.target, residual=res, jacobian=jac, lhs0=lhs, constrain=self.constrain,
-        arguments=collections.ChainMap(self.arguments, {self.target0: lhs}), **self.newtonargs).solve(tol=self.newtontol)
+        arguments=collections.ChainMap(self.arguments, {self.target0: lhs, self.timetarget: t}), **self.newtonargs).solve(tol=self.newtontol)
     except (SolverError, matrix.MatrixError) as e:
       log.error('error: {}; retrying with timestep {}'.format(e, timestep/2))
-      return self._step(self._step(lhs, timestep/2), timestep/2)
+      return self._step(self._step(lhs, t, timestep/2), t+timestep/2, timestep/2)
 
   def resume(self, history):
     if history:
-      lhs, = history
+      (lhs, i), = history
     else:
-      lhs = self.lhs0
-      yield lhs
+      lhs, i = self.lhs0, 0
+      yield lhs, i
     while True:
-      lhs = self._step(lhs, self.timestep)
+      lhs = self._step(lhs, self.time0+i*self.timestep, self.timestep)
+      i += 1
+      yield lhs, i
+
+  def __iter__(self):
+    for lhs, i in super().__iter__():
       yield lhs
```

Comment:
Isn't it cleaner to have "i" be retrievable from the cacher, rather than caching the enumeration and then removing it? After all the cacher has this number, it just doesn't currently volunteer it. Not sure if that can be done in a backwards-compatible way though so that might argue against it.

Reply:
I think the only way we could do this is by adding an argument to the `resume` method, which breaks existing subclasses (outside Nutils). We could of course support a resume method with both one and two arguments (besides `self`) via introspection, and issue a deprecation warning in case of one argument.

Reply:
PR #417 seems to be a fairly clean way of doing this?

Reply:
Indeed!

Reply:
Merged.